src/app/server-cluster/AttributeListBuilder.cpp

```diff
@@ -22,13 +22,14 @@ namespace chip {
 namespace app {
 
 CHIP_ERROR AttributeListBuilder::Append(Span<const DataModel::AttributeEntry> mandatoryAttributes,
-                                        Span<const OptionalAttributeEntry> optionalAttributes)
+                                        Span<const DataModel::AttributeEntry> optionalAttributes,
```

Comment:
As a followup, if we now take this approach, then we can codegen the optionalAttributes as metadata too, right?

Reply:
Yes, for the most common case. We have a few cases where these do not apply:

- UniqueID for BasicInfo needs to be optional even if mandatory in the spec (because we need code to deny it)
- UpTime in generaldiagnostics is optional in the XML, but mandatory in the spec (which is different from what we did with uniqueid ...).

I would say we should generate these and then manually adjust for these odd corner cases. We should also standardize what we actually do in the XML for things that were optional but became mandatory.

Reply:
We also probably cannot fully generalize this as long as we have a bitset restriction: we have a few clusters with attribute IDs > 31 or even 63 and we have a few clusters with a lot of attributes (70+).

This would work for most clusters, but not all and a few may need special casing.